sdk/eventhubs/azure_messaging_eventhubs/src/common/recoverable/connection.rs

```diff
@@ -0,0 +1,473 @@
+    ///
+    /// # Note
+    ///
+    /// This method is public(crate) to allow event producers and event consumers to
+    /// verify that the underlying connection is established before finishing the
+    /// construction of the underlying client - this avoids the "magic function" problem
+    /// where the client is constructed, but the connection is not established until the
+    /// first operation is performed.
+    ///
+    pub(crate) async fn ensure_connection(&self) -> Result<Arc<AmqpConnection>> {
+        let connection = self
+            .connections
+            .get_or_try_init(|| self.create_connection())
+            .await?;
+        Ok(connection.clone())
+    }
+
+    /// Creates a new management client for the Event Hubs service.
+    ///
+    /// This client is used to perform management operations such as querying the status of the Event Hubs service.
+    pub(crate) async fn get_management_client(
+        self: &Arc<Self>,
+    ) -> Result<RecoverableManagementClient> {
+        Ok(RecoverableManagementClient::new(self.clone()))
+    }
+
+    /// Creates a new Claims-Based Security (CBS) client for the Event Hubs service.
+    ///
+    /// This client is used to perform authorization operations such as acquiring tokens for accessing Event Hubs resources.
+    ///
+    /// Note: The Cbs client returned integrates retry operations into the authorization call.
+    pub(crate) async fn get_cbs_client(self: &Arc<Self>) -> Result<RecoverableClaimsBasedSecurity> {
+        Ok(RecoverableClaimsBasedSecurity::new(self.clone()))
+    }
+
+    /// Creates a new sender for the Event Hubs service.
+    ///
+    /// # Notes
+    ///
+    /// This sender integrates retry operations into the send operation.
+    pub(crate) async fn get_sender(self: &Arc<Self>, path: Url) -> Result<RecoverableSender> {
+        // Ensure we can create a sender for the Event Hub path.
+        self.ensure_sender(&path).await?;
+
+        Ok(RecoverableSender::new(self.clone(), path))
+    }
+
+    pub(crate) async fn get_receiver(
+        self: &Arc<Self>,
+        source_url: &Url,
+        message_source: AmqpSource,
+        receiver_options: AmqpReceiverOptions,
+        timeout: Option<Duration>,
+    ) -> Result<RecoverableReceiver> {
+        self.ensure_receiver(source_url, &message_source, &receiver_options)
+            .await?;
+
+        Ok(RecoverableReceiver::new(
+            self.clone(),
+            receiver_options,
+            message_source,
+            source_url.clone(),
+            timeout,
+        ))
+    }
+
+    pub(crate) async fn close_receiver(self: &Arc<Self>, source_url: &Url) -> Result<()> {
+        let mut receiver_instances = self.receiver_instances.lock().await;
+        if let Some(receiver) = receiver_instances.remove(source_url) {
+            let r = Arc::try_unwrap(receiver);
+            if let Ok(receiver) = r {
+                trace!("Detaching receiver: {:?}", source_url);
```

Comment:
Tip: only because you inline some stuff elsewhere, in case you weren't aware, you can also combine these:
```suggestion
                trace!("Detaching receiver: {source_url:?}");
```
You don't have to change this, though. Just a tip.